app/upgrades/v17/upgrades_test.go

```diff
@@ -137,12 +205,96 @@ func (suite *UpgradeTestSuite) TestUpgrade() {
+					suite.Require().Equal(clPool1TwapRecordHistoricalPoolIndexPreUpgrade[i].P1LastSpotPrice, clPool1TwapRecordHistoricalPoolIndexPostUpgrade[i].P0LastSpotPrice)
+				}
+
+				// check that all most recent TWAP record has been flipped for pool2
+				for i := range clPool2TwapRecordPostUpgrade {
+					suite.Require().Equal(clPool2TwapRecordPreUpgrade[i].Asset0Denom, clPool2TwapRecordPostUpgrade[i].Asset1Denom)
+					suite.Require().Equal(clPool2TwapRecordPreUpgrade[i].Asset1Denom, clPool2TwapRecordPostUpgrade[i].Asset0Denom)
+					suite.Require().Equal(clPool2TwapRecordPreUpgrade[i].P0LastSpotPrice, clPool2TwapRecordPostUpgrade[i].P1LastSpotPrice)
+					suite.Require().Equal(clPool2TwapRecordPreUpgrade[i].P1LastSpotPrice, clPool2TwapRecordPostUpgrade[i].P0LastSpotPrice)
+				}
+
+				// check that historical pool index TWAP record has been flipped for pool2
+				for i := range clPool2TwapRecordHistoricalPoolIndexPostUpgrade {
+					suite.Require().Equal(clPool2TwapRecordHistoricalPoolIndexPreUpgrade[i].Asset0Denom, clPool2TwapRecordHistoricalPoolIndexPostUpgrade[i].Asset1Denom)
+					suite.Require().Equal(clPool2TwapRecordHistoricalPoolIndexPreUpgrade[i].Asset1Denom, clPool2TwapRecordHistoricalPoolIndexPostUpgrade[i].Asset0Denom)
+					suite.Require().Equal(clPool2TwapRecordHistoricalPoolIndexPreUpgrade[i].P0LastSpotPrice, clPool2TwapRecordHistoricalPoolIndexPostUpgrade[i].P1LastSpotPrice)
+					suite.Require().Equal(clPool2TwapRecordHistoricalPoolIndexPreUpgrade[i].P1LastSpotPrice, clPool2TwapRecordHistoricalPoolIndexPostUpgrade[i].P0LastSpotPrice)
+				}
+
+				// check that historical time index TWAP record has been flipped for pool1 and pool2
+				// Also check that all the records is the same for the rest of the pools
+				for i := range clPoolsTwapRecordHistoricalTimeIndexPostUpgrade {
+					if (clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].PoolId == lastPoolIdMinusOne) || (clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].PoolId == lastPoolIdMinusTwo) {
+						suite.Require().Equal(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i].Asset0Denom, clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].Asset1Denom)
+						suite.Require().Equal(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i].Asset1Denom, clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].Asset0Denom)
+						suite.Require().Equal(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i].P0LastSpotPrice, clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].P1LastSpotPrice)
+						suite.Require().Equal(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i].P1LastSpotPrice, clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].P0LastSpotPrice)
+					} else {
+						if clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i].PoolId == lastPoolID {
+							suite.Require().Equal(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i], clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i])
+						}
+					}
+				}
```

Comment:
I think all of this code can be abstracted into the following:

```
func assertTwapFlipped(pre, post interface{}) {
	suite.Require().Equal(pre.Asset0Denom, post.Asset1Denom)
	suite.Require().Equal(pre.Asset1Denom, post.Asset0Denom)
	suite.Require().Equal(pre.P0LastSpotPrice, post.P1LastSpotPrice)
	suite.Require().Equal(pre.P1LastSpotPrice, post.P0LastSpotPrice)
}

func assertEqual(pre, post interface{}) {
	suite.Require().Equal(pre, post)
}

for _, data := range []struct {
	pre, post interface{}
}{
	{clPool1TwapRecordPreUpgrade, clPool1TwapRecordPostUpgrade},
	{clPool1TwapRecordHistoricalPoolIndexPreUpgrade, clPool1TwapRecordHistoricalPoolIndexPostUpgrade},
	{clPool2TwapRecordPreUpgrade, clPool2TwapRecordPostUpgrade},
	{clPool2TwapRecordHistoricalPoolIndexPreUpgrade, clPool2TwapRecordHistoricalPoolIndexPostUpgrade},
} {
	for i := range data.post {
		assertTwapFlipped(data.pre[i], data.post[i])
	}
}

for i := range clPoolsTwapRecordHistoricalTimeIndexPostUpgrade {
	record := clPoolsTwapRecordHistoricalTimeIndexPostUpgrade[i]
	if record.PoolId == lastPoolIdMinusOne || record.PoolId == lastPoolIdMinusTwo {
		assertTwapFlipped(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i], record)
	} else if record.PoolId == lastPoolID {
		assertEqual(clPoolsTwapRecordHistoricalTimeIndexPreUpgrade[i], record)
	}
}

```

Reply:
this is gold